EyeSeeTea/glass-dev
# Script AMC recalculations

### :pushpin: References

-   **Issue:** Closes [AMC script for recalculations](https://app.clickup.com/t/8693ecv17)

### :memo: Implementation
- [x] Create script to allow recalculate all stored data of AMC product level and AMC substance level using current ATC data version every time in **datastore --> glass --> amc-recalculation** the value `recalculate` is true (using the organization units and periods added here)
- [x] Allow to build script for running it in background of DHIS2 server (as [metadata-synchronization scheduler](https://github.com/EyeSeeTea/metadata-synchronization))
- [x] Allow to create events with calculated data if current don't exist when flag `--calculate` is true 

### :video_camera: Screenshots/Screen capture
![Screenshot from 2024-01-19 10-38-40](https://github.com/EyeSeeTea/glass-dev/assets/49402898/95df2f1b-42e7-46ce-9326-82655e57966f)
![Screenshot from 2024-01-19 10-38-45](https://github.com/EyeSeeTea/glass-dev/assets/49402898/a25f3883-e38c-44bb-8607-eab95b7ca5f6)

![Screenshot from 2024-01-19 10-40-18](https://github.com/EyeSeeTea/glass-dev/assets/49402898/040eaf81-c15a-4310-baa4-7865a0a29c51)
![Screenshot from 2024-01-19 10-40-26](https://github.com/EyeSeeTea/glass-dev/assets/49402898/a296ff4d-2485-45f2-abf9-5942a1dd316d)


### :fire: Testing
1. Product Level data: in tracker capture check in Thailand in program AMC - Product Register a product with enrollment date in 2020 period, and check inside the values of the events of stage AMC - Raw Substance Consumption Calculated
2. Substance level data: In Capture in Thailand in AMC - Calculated Consumption Data program check events with report date in period 2020, open one and check values
3. Change current ATC version in datastore glass --> ATCs 
4. In datastore have in glass --> amc-recalculation:
```
{
  "date": "2024-01-19T09:54:04.091Z",
  "orgUnitsIds": [
    "vboedbUs1As"
  ],
  "periods": [
    "2020"
  ],
  "recalculate": false
}
``` 
And change recalculate to true
5. Start script: 
`yarn start-amc-recalculate --url "https://dev.eyeseetea.com/who-dev-238" --auth USERNAME:PASSWORD`
Flag to also create events with calculations if it does not exist: `yarn start-amc-recalculate --url "https://dev.eyeseetea.com/who-dev-238" --auth USERNAME:PASSWORD --calculate`
7. After the script ends, check again in Tracker Capture and in Capture if that events change ATC version to current one. In real life the ATC data of this version should change so also the values in the events will change
